packages/ra-ui-materialui/src/input/DateInput.js

```diff
@@ -23,59 +23,69 @@ const dateFormatter = v => {
 
 const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
 
-const sanitizeValue = value => {
+const format = value => {
     // null, undefined and empty string values should not go through dateFormatter
     // otherwise, it returns undefined and will make the input an uncontrolled one.
     if (value == null || value === '') {
         return '';
     }
 
     if (value instanceof Date) {
-        return dateFormatter(value);
+        return convertDateToString(value);
     }
 
     // valid dates should not be converted
     if (dateRegex.test(value)) {
         return value;
     }
 
-    return dateFormatter(new Date(value));
+    return convertDateToString(new Date(value));
 };
 
+/**
+ * Converts a datetime string without timezone to a date object
+ * with timezone, using the browser timezone.
+ *
+ * @param {String} value Date string, formatted as yyyy-MM-ddThh:mm
+ * @return {Date}
+ */
+const parse = value => (value ? new Date(value) : '');
```

Comment:
I'm still bothered by this timezone cast..
If we don't have a timezone here, it's because the `format` stripped it.

IMO, `convertDateToString` should just use `date.toISOString()` so that we keep the right timezone all the way.

Reply:
As I said before, I'm reluctant to change the input behavior in this PR. Please open an issue about it so that we don't forget though

Reply:
Ok, I checked that out before opening a PR, and I was wrong.

This input type has two drawbacks:
- It is not supported by some major browsers used now (Firefox and Safari) #2685
- It format the date using the non-standard `yyyy-MM-ddThh:mm` format instead of [ISO 8601 Extended Format](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Date/toISOString) (eg. `YYYY-MM-DDTHH:mm:ss.sssZ`). Removing support for seconds and timezone.

But, the [`datetime` input type is deprecated by the WHATWG HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime).
Moreover, [no other date or time date is supported by Firefox and Safari](https://caniuse.com/#feat=input-datetime).

Hence, the `datetime-local` seems to be the right input type choice if we don't want to use external library to manage this input.

Now one question: should the `<DateTimeInput />` return a string with the format  `yyyy-MM-ddThh:mm` or a `Date` with the current user timezone and seconds set to 00?

Reply:
This PR only deals with the DateInput. I opened another one for the DateTimeInput

Reply:
Besides, as said before, I don't want to change its behavior here. Just convert it to hooks. Please open an issue about date/time handling if needed

Reply:
Sorry, I'm not sure to understand then.
By passing a `parse` function to the `useInput` you change the behavior: DateInput returns now a Date instead of a string.
Or am I missing something?

Reply:
You're right, I'll remove it for now